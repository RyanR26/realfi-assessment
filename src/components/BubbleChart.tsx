import React, { useMemo, useRef } from 'react';
import counter from '../utils/counter';

interface IProps {
  dataPoints: IDataPoint[],
  config?: {
    title?: string,
    tooltipLabel?: string,
    height?: string,
    width?: string,
    containerPadding?: string,
    groupClasses?: string[]
  }
}

interface IDataPoint { 
  value: number, 
  label: string,
  groupClass?: string
}

function getSqRootValue (a: number, b: number): number {
  return Math.sqrt(((a/b) * 100) * 2)
};

export default function BubbleChart(props: IProps) {

  const { 
    dataPoints,
    config 
  } = props;

  const dataPointsSorted = useMemo(() => dataPoints.sort(function (a, b) {  return b.value - a.value }), [dataPoints]);
  const maxValue = dataPointsSorted[0].value;
  const maxSqrValue = useMemo(() => getSqRootValue(maxValue, maxValue), [maxValue]);
  const height = config?.height || '500px';
  const width = config?.width || '100%';
  const containerPadding = config?.containerPadding || '10px';
  const countDivisor = config?.groupClasses?.length || 0;
  const counterRef = useRef(counter(countDivisor));

  return (
    <div className='bubble-chart'>
      { config?.title && <div className='text-md text-center pb-4 underline'>{ config.title }</div> }
      <div className='bubbles-container rounded-box border-2 border-primary p-4 relative' style={{ height, width }}>
      {
        dataPointsSorted.map((dataPoint: IDataPoint, index: number) => {

          if (dataPoint.value > 0) {
            return (    
              <div 
                key={dataPoint.label} 
                className={`${config?.groupClasses ? config.groupClasses[counterRef.current(index)] : 'bg-secondary'} mt-auto mx-auto mb-0 w-1/2 flex justify-center text-black absolute rounded-xl pt-4 tooltip`}
                data-tip={`${dataPoint.value} ${config?.tooltipLabel}`}
                style={{ 
                  height: `calc(${(getSqRootValue(dataPoint.value, maxValue)/maxSqrValue) * 100}% - ${containerPadding})`,
                  width: `calc(${(getSqRootValue(dataPoint.value, maxValue)/maxSqrValue) * 100}% - ${containerPadding})`, 
                  borderRadius: '100%',
                  bottom: `calc(${containerPadding}/2)`,
                  border: config?.groupClasses ? '' : '1px dashed grey',
                  left: '50%',
                  transform: 'translateX(-50%)',
                  transition: 'all 500ms ease'
                }}>
                  <span className='text-xs'>{dataPoint.label} - {dataPoint.value}</span>
                </div>)
          } else {
            return null
          }
        })
      }
      </div>
    </div>
  )
};