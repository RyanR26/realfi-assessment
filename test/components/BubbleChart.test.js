import renderer from 'react-test-renderer'
import BubbleChart from '../../src/components/BubbleChart'

it('Should render the BubbleChart HMTL correctly against the snapshot', () => {

  const bubbleChartDataPoints = [
    {
        "value": 9,
        "label": "Latvia"
    },
    {
        "value": 2,
        "label": "Germany"
    },
    {
        "value": 1,
        "label": "China"
    },
    {
        "value": 0,
        "label": "Spain"
    }
  ];

  const config = {
    "title": "Dependents by country",
    "tooltipLabel": "Dependant(s)",
    "height": "300px",
    "width": "300px",
    "containerPadding": "40px",
    "groupClasses": [
        "bg-primary",
        "bg-secondary",
        "bg-accent"
    ]
  };

  const component = renderer.create(<BubbleChart dataPoints={bubbleChartDataPoints} config={config}/ >);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  {/* // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseEnter();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  // manually trigger the callback
  renderer.act(() => {
    tree.props.onMouseLeave();
  });
  // re-rendering
  tree = component.toJSON();
  expect(tree).toMatchSnapshot(); */}
});