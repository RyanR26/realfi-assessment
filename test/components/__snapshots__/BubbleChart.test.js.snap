// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render the BubbleChart HMTL correctly against the snapshot 1`] = `
<div
  className="bubble-chart"
>
  <div
    className="text-md text-center pb-4 underline"
  >
    Dependents by country
  </div>
  <div
    className="bubbles-container rounded-box border-2 border-primary p-4 relative"
    style={
      {
        "height": "300px",
        "width": "300px",
      }
    }
  >
    <div
      className="bg-primary mt-auto mx-auto mb-0 w-1/2 flex justify-center text-black absolute rounded-xl pt-4 tooltip"
      data-tip="9 Dependant(s)"
      style={
        {
          "border": "",
          "borderRadius": "100%",
          "bottom": "calc(40px/2)",
          "height": "calc(100% - 40px)",
          "left": "50%",
          "transform": "translateX(-50%)",
          "transition": "all 500ms ease",
          "width": "calc(100% - 40px)",
        }
      }
    >
      <span
        className="text-xs"
      >
        Latvia
         - 
        9
      </span>
    </div>
    <div
      className="bg-secondary mt-auto mx-auto mb-0 w-1/2 flex justify-center text-black absolute rounded-xl pt-4 tooltip"
      data-tip="2 Dependant(s)"
      style={
        {
          "border": "",
          "borderRadius": "100%",
          "bottom": "calc(40px/2)",
          "height": "calc(47.14045207910317% - 40px)",
          "left": "50%",
          "transform": "translateX(-50%)",
          "transition": "all 500ms ease",
          "width": "calc(47.14045207910317% - 40px)",
        }
      }
    >
      <span
        className="text-xs"
      >
        Germany
         - 
        2
      </span>
    </div>
    <div
      className="bg-accent mt-auto mx-auto mb-0 w-1/2 flex justify-center text-black absolute rounded-xl pt-4 tooltip"
      data-tip="1 Dependant(s)"
      style={
        {
          "border": "",
          "borderRadius": "100%",
          "bottom": "calc(40px/2)",
          "height": "calc(33.33333333333333% - 40px)",
          "left": "50%",
          "transform": "translateX(-50%)",
          "transition": "all 500ms ease",
          "width": "calc(33.33333333333333% - 40px)",
        }
      }
    >
      <span
        className="text-xs"
      >
        China
         - 
        1
      </span>
    </div>
  </div>
</div>
`;
